### PR TITLE
Add HIDDEN_FOLDER_NAMES class to shares.py for easier management to exclude multiple folders from shares

### PR DESCRIPTION
**Pull request by toineenzo**
Added the Synology #snapshot and #recycle folders to be excluded from Share as these folders contain data that have been already deleted which may be a security issues and may flood the SLSK network with unnecessary files.

### CODE REVIEW

File: pynicotine/shares.py [hunk 1]
```diff
@@ -477,11 +477,11 @@ def is_hidden(folder, filename=None, entry=None):
         """Stop sharing any dot/hidden folders/files."""
 
         # If the last folder in the path starts with a dot, or is a Synology extended
-        # attribute folder, we exclude it
+        # attribute, snapshot or recycle bin folder, we exclude it
         if filename is None:
             last_folder = os.path.basename(folder)
 
-            if last_folder.startswith(".") or last_folder == "@eaDir":
+            if last_folder.startswith(".") or last_folder == ["@eaDir", "#recycle", "#snapshot"]:
```
Comment (slook) [inline pynicotine/shares.py:484]: Condition is equal to entire list. You need `in` instead of `==`.

**Review by slook — CHANGES_REQUESTED** [review]

File: pynicotine/shares.py (hunk 1 above)
Comment (toineenzo) [reply pynicotine/shares.py]: Thank you, fixed.

**Comment by slook** [comment]
https://github.com/nicotine-plus/nicotine-plus/actions/runs/7718684220/job/21040477536?pr=2843#step:6:169
> `pynicotine/shares.py:484:61: R6201: Consider using set for membership test (use-set-for-membership)`

To use a set is as simple as using `{`curly brackets`}` in place of `[`square brackets`]` to contain the list of strings.

File: pynicotine/shares.py [hunk 2]
```diff
@@ -477,11 +477,11 @@ def is_hidden(folder, filename=None, entry=None):
         """Stop sharing any dot/hidden folders/files."""
 
         # If the last folder in the path starts with a dot, or is a Synology extended
-        # attribute folder, we exclude it
+        # attribute, snapshot or recycle bin folder, we exclude it
         if filename is None:
             last_folder = os.path.basename(folder)
 
-            if last_folder.startswith(".") or last_folder == "@eaDir":
+            if last_folder.startswith(".") or last_folder in ["@eaDir", "#recycle", "#snapshot"]:
```
Comment (mathiascode) [inline pynicotine/shares.py:484]: I would make the folder list a class constant, such as:
`HIDDEN_FOLDER_NAMES = {"@eaDir", "#recycle", "#snapshot"}`

then access it with

`if last_folder.startswith(".") or last_folder in self.HIDDEN_FOLDER_NAMES:`

Comment (toineenzo) [reply pynicotine/shares.py]: Done

File: pynicotine/shares.py [hunk 3]
```diff
@@ -477,14 +477,15 @@ def is_hidden(folder, filename=None, entry=None):
         """Stop sharing any dot/hidden folders/files."""
 
         # If the last folder in the path starts with a dot, or is a Synology extended
-        # attribute folder, we exclude it
+        # attribute, snapshot or recycle bin folder, we exclude it
         if filename is None:
             last_folder = os.path.basename(folder)
 
-            if last_folder.startswith(".") or last_folder == "@eaDir":
+            HIDDEN_FOLDER_NAMES = {"@eaDir", "#recycle", "#snapshot"}
```
Comment (mathiascode) [inline pynicotine/shares.py:484]: Please move this to the beginning of the class, below the initial comment (line 232).

File: pynicotine/shares.py [hunk 4]
```diff
@@ -477,14 +477,15 @@ def is_hidden(folder, filename=None, entry=None):
         """Stop sharing any dot/hidden folders/files."""
 
         # If the last folder in the path starts with a dot, or is a Synology extended
-        # attribute folder, we exclude it
+        # attribute, snapshot or recycle bin folder, we exclude it
```
Comment (mathiascode) [inline pynicotine/shares.py:480]: This comment should preferably be more generic, so you don't have to update it if the list of hidden folders is updated in the future.

Perhaps something like this:

> If the last folder in the path starts with a dot, or uses the same name as well-known hidden folders, we exclude it

**Comment by mathiascode** [comment]
I forgot the is_hidden() method is static. Change the function header to:

```
    @classmethod
    def is_hidden(cls, folder, filename=None, entry=None):
```

and `self.HIDDEN_FOLDER_NAMES` to `cls.HIDDEN_FOLDER_NAMES`

File: pynicotine/shares.py (hunk 2 above)
Comment (mathiascode) [reply pynicotine/shares.py]: No worries!

**Comment by slook** [comment]
It's too late to implement this new feature whilst in release candidate stage, so I propose we pause this PR until the next development cycle in 3.3.1dev1.

File: pynicotine/shares.py (hunk 3 above)
Comment (toineenzo) [reply pynicotine/shares.py:484]: I've moved it below this:
```
class Scanner:
    """Separate process responsible for building shares.

    It handles scanning of folders and files, as well as building
    databases and writing them to disk.
    """
```

**Comment by toineenzo** [comment]
> It's too late to implement this new feature whilst in release candidate stage, so I propose we shouldn't merge this PR until the next development cycle in 3.3.1dev1.



> It's too late to implement this new feature whilst in release candidate stage, so I propose we shouldn't merge this PR until the next development cycle in 3.3.1dev1.

Will the excluding of the @eaDir folder still be included in 3.3.0 stable as currently in the latest 3.3.0-rc3?

**Comment by toineenzo** [comment]
> It's too late to implement this new feature whilst in release candidate stage, so I propose we shouldn't merge this PR until the next development cycle in 3.3.1dev1.

Will the excluding of the @ eadir folder still be included in 3.3.0 stable as currently in the latest 3.3.0-rc3?

File: pynicotine/shares.py [hunk 5]
```diff
@@ -472,19 +475,19 @@ def rescan_dirs(self, permission_level):
...
-    @staticmethod
-    def is_hidden(folder, filename=None, entry=None):
+    @classmethod
+    def is_hidden(cls, folder, filename=None, entry=None):
         """Stop sharing any dot/hidden folders/files."""
 
-        # If the last folder in the path starts with a dot, or is a Synology extended
-        # attribute folder, we exclude it
+        # If the last folder in the path starts with a dot, or uses
+        # the same name as well-known hidden folders, we exclude it
```
Comment (slook) [inline pynicotine/shares.py:483]: Remove this entire comment, the code is self-explainatory.

File: pynicotine/shares.py [hunk 6]
```diff
@@ -231,6 +231,9 @@ class Scanner:
     databases and writing them to disk.
     """
 
+    # Add foldernames that need to be excluded from shares
```
Comment (slook) [inline pynicotine/shares.py:234]: Remove this entire comment, the code is self-explainatory.

**Comment by slook** [comment]
> Will the excluding of the @ eadir folder still be included in 3.3.0 stable as currently in the latest 3.3.0-rc3?

Yes.

**Review by mathiascode — APPROVED** [review]

**Pull request merged**